private/bufpkg/bufmodule/module.go

```diff
@@ -452,33 +452,31 @@ func newGetDigestFuncForModuleAndDigestType(module *module, digestType DigestTyp
 			}
 			return getB4Digest(module.ctx, bucket, v1BufYAMLObjectData, v1BufLockObjectData)
 		case DigestTypeB5:
-			moduleDeps, err := module.ModuleDeps()
-			if err != nil {
-				return nil, err
-			}
-			// For remote modules to have consistent B5 digests, they must not change the digests of their
-			// dependencies based on the local workspace. Use the pruned b5 module keys from
-			// ModuleData.DeclaredDepModuleKeys to calculate the digest.
+			// B5 digests are calculated from the Module's content and its dependencies.
+			//
+			// Dependencies are all direct and transitive dependencies of the Module.
+			// For local Modules, the dependencies are resolved to the current ModuleSet.
+			// For remote Modules, the dependencies are the ModuleKeys of
+			// the direct and transitive dependencies of the Module at the specific commit.
+			//
+			// This is effectively the same. The local Modules have a dependency at the current commit,
+			// and the remote Modules have a dependency at a specific commit. Pushing up a local Module
+			// as a remote Module will result in the same Digest.
+			//
+			// See the comment on getDepModuleKeysB5 for more information.
 			if !module.isLocal {
-				declaredDepModuleKeys, err := module.getDeclaredDepModuleKeysB5()
+				depModuleKeys, err := module.getDepModuleKeysB5()
 				if err != nil {
 					return nil, err
 				}
-				moduleDepFullNames := make(map[string]struct{}, len(moduleDeps))
-				for _, dep := range moduleDeps {
-					fullName := dep.FullName()
-					if fullName == nil {
-						return nil, syserror.Newf("remote module dependencies should have full names")
-					}
-					moduleDepFullNames[fullName.String()] = struct{}{}
-				}
-				prunedDepModuleKeys := make([]ModuleKey, 0, len(declaredDepModuleKeys))
```

Comment:
This is the fix and the cause of the tamper-proof error. Dependencies should **not** be filtered to the context of the parent Module. This is required to avoid the Digest for the remote changing when included in a Module. If the parent Module `A` doesn't include dependency `X` but the remote module `B` depends on `X`, its dependency should still be used in the calculation of the Digest for `B`.